File: renderer/Scene/PbrScene.cpp
#include "PbrScene.h"
#include <memory>
#include <assert.h>
#include <algorithm>
#include "imgui/imgui_impl_win32.h"
#include "imgui/imgui.h"
#include "Math/Util.h"
#include "Render/Graph/PassNode.h"
#include "Render/Graph/ResourceEntry.h"
#include "Render/Mesh/model.h"
#include "render/camera.h"
#include "Math/Vec3.h"
#include "Math/Mat4.h"
#include "Core/GameObject.h"
#include "Behaviour/CameraController.h"
#include "Render/Renderer.h"
#include "Render/Light.h"
#include "Render/Mesh/Primitive.h"
#include "Render/Base/CommandQueue.h"

namespace glacier {
namespace render {

static void DrawHelmet(CommandBuffer* cmd_buffer) {
    auto& helmet = Model::GenerateGameObject(cmd_buffer, "assets\\model\\helmet\\DamagedHelmet.gltf", true, 1.0f);
    //helmet.transform().position(Vector3{ 6.0f,4.0f,10.0f });
    helmet.transform().position(Vector3{ 99, 25, 4 });
}

static void DrawSponza(CommandBuffer* cmd_buffer) {
    auto& sponza = Model::GenerateGameObject(cmd_buffer, "assets\\model\\sponza\\sponza.gltf", true, 10.0f);
    //sponza.transform().position({ 6.0f,4.0f,10.0f });
}

void PbrScene::OnLoad(Renderer* renderer) {
    auto gfx = renderer->driver();

    auto& camera_go = GameObject::Create("MainCamera");
    main_camera_ = camera_go.AddComponent<Camera>();
    main_camera_->position({ 102.0f, 25.5f, 4.5f });

    auto rot = Quaternion::FromEuler(9.4f, 266.671f, 0);
    main_camera_->rotation(rot);

    camera_go.AddComponent<CameraController>(main_camera_);

    auto& light_go = GameObject::Create("main_light");
    auto dir = (Vec3f(1, -1, 0) - Vec3f::zero).Normalized();
    light_go.transform().forward(dir);

    auto main_light = light_go.AddComponent<DirectionalLight>(Color::kWhite, 1.0f);
    main_light->EnableShadow();

    auto cmd_queue = gfx->GetCommandQueue(CommandBufferType::kDirect);
    auto cmd_buffer = cmd_queue->GetCommandBuffer();

    //DrawHelmet(cmd_buffer);
    DrawSponza(cmd_buffer);

    cmd_queue->ExecuteCommandBuffer(cmd_buffer);
    cmd_queue->Flush();
}

}
}

File: renderer/Script/pbrscene.lua
local Scene = require("Glacier.Scene")
local GameObject = require("Glacier.GameObject")
local Camera = require("Glacier.Camera")
local CameraType = require("Glacier.CameraType")
local Quaternion = require("core.math.quaternion")
local Vector3 = require("core.math.vector3")
local CameraController = require("Glacier.CameraController")
local DirectionalLight = require("Glacier.DirectionalLight")
local CommandBufferType = require("Glacier.CommandBufferType")
local Model = require("Glacier.Model")
local MaterialManager = require("Glacier.MaterialManager")
local Primitive = require("Glacier.Primitive")

local pbr_scene = Scene("pbr")

pbr_scene:OnLoad(function(renderer)
    local camera_go = GameObject.Create("MainCamera"):gc_disable()
    local camera = Camera(CameraType.kPersp):gc_disable()
    pbr_scene:SetMainCamera(camera)

    camera_go:AddComponentPtr(camera)
    local cam_ctrl = CameraController(camera):gc_disable()
    camera_go:AddComponentPtr(cam_ctrl)

    camera.position = { 102.0, 25.5, 4.5 }
    camera.rotation = Quaternion.euler(9.4, 266.671, 0)

    local light_go = GameObject.Create("main_light"):gc_disable()
    local dir = (Vector3(1, -1, 0) - Vector3.zero):normalize()
    light_go:GetTransform().forward = dir;

    local main_light = DirectionalLight({1, 1, 1, 1}, 1.0):gc_disable()
    main_light:EnableShadow()
    light_go:AddComponentPtr(main_light);

    local gfx = renderer:driver()
    local cmd_queue = gfx:LGetCommandQueue(CommandBufferType.kDirect)
    local cmd_buffer = cmd_queue:GetCommandBuffer()

    local helmet = Model.GenerateGameObject(cmd_buffer, "assets\\model\\helmet\\DamagedHelmet.gltf", true, 1.0):gc_disable()
    helmet:GetTransform().position = { 99, 25, 4 }

    local pbr_floor = MaterialManager.Instance():Get("pbr_floor");

    local ground_go = Primitive.CreateCube(pbr_floor, {200.0, 1.0, 200.0}):gc_disable()
    ground_go.name = "Ground"
    ground_go:GetTransform().position = { 0.0, -1.5, 0.0 }

    cmd_queue:LExecuteCommandBuffer(cmd_buffer)
    cmd_queue:LFlush()
end)

return pbr_scene
